### ISSUE TITLE
Change hashing scheme and remove need for ordering

### ISSUE THREAD

**Issue by keatonwalker** [agrc/forklift#152]
### Changes to improve hashing scheme:
1. **Combine attribute and geometry hashes**
    1. Size of hash table in memory and on disk would be smaller
    1.  Uniqueness of the hashes would be improved
1. **Add salt only to identical rows**
    1. No need for ordering
    1. Salt could be the hash of the previous identical row
    1. The total number of identical rows could be maintained in source and destination
     1. Any row of an identical group may be removed from destination when one is removed from source
1. **Try xxhash**
    1. Likely much faster than md5
    1. Hex digests are half the size of md5
    1. Code changes would be very easy (1 or 2 lines)

**Comment by stdavis** [comment]
1 & 3 sound great to me.

I'm still struggling to catch up on 2. I read the thread in slack but still not quite there. Can you clarify what it means to maintain the total number of identical rows in the source and destination?

**Comment by keatonwalker** [comment]
I'm sure you already got some of this from the slack thread but I will try to summarize it here, for posterity.

**2.iii The total number of identical rows would be maintained in source and destination**
Identical rows are rows thats have the same attributes and same geometry. A list representation of identical row tuples with fields `(OID, field1, field2, field3, shape_x, shape_y)` would be:
```
identical_grp = [(1, 1.234, 'text stuff', 9, 3000.345, 4000.456),
                 (2, 1.234, 'text stuff', 9, 3000.345, 4000.456),
                 (3, 1.234, 'text stuff', 9, 3000.345, 4000.456)]
```
If row 1, 2 or 3 were hashed without any salt, and ignoring the `OID` field, they would all hash to the same value. Let's say that value is `'abc'`. So 2.iii above would mean that if there are 3 rows that would hash to `'abc'` without salt in source, there will be 3 rows that hash to `'abc'` in destination.

**2.iv Any row of an identical group may be removed from destination when one is removed from source**
Adding salt to only the identical rows allows us to keep track of the total number of identical rows in source and destination, but not the exact rows that might be added or deleted. 
- First, a few definitions to make typing this easier:
`h(identical_grp[n]) = 'abc'` where `h()` is the hash function and `n` is any index in `identical_grp`.
- Using the hash of the previous row in an identical group as salt (2.ii) would look like this:
The first row's hash would be `h(identical_grp[0]) = 'abc'`
The second row's hash would then be `h(identical_grp[1] + h(identical_grp[0])) = 'aaa'`
And the third row's `h(identical_grp[2] + h(identical_grp[1] + h(identical_grp[0])) = 'bcb'`
- So let's say the OID and hash associations  for source and destination are:

| Source OID  | Hash | Destination OID |
| ------------- | ------------- | ------------- |
| 1 | 'abc'  | 4 |
| 2  | 'aaa'  | 5 |
| 3  | 'bcb'  | 6 |

- Now OID 2 gets deleted in source, and after another run, the association table would be:

| Source OID  | Hash | Destination OID |
| ------------- | ------------- | ------------- |
| 1 | 'abc'  | 4 |
| 3  | 'aaa'  | 5 |

  and OID 6 would be deleted from destination.

The bottom line concerning 2.iv is that there is no useful association between specific identical rows in source and destination. We can only know that the number of rows that hash to `'abc'` without salt will always be the same in both places.

**Comment by stdavis** [comment]
Thanks for the clarification. I think that I'm slowly coming along. I agree that it's not essential to maintain a one-to-one relationship of features within groups of identical rows between source and destination as long as the number of them is the same in each dataset. 👍

**Comment by steveoh** [comment]
This would remove duplicate records in the destination if I am following along. So the number of items would not be equal in each dataset. Can you validate that assumption @keatonwalker

**Comment by keatonwalker** [comment]
@steveoh , no it wouldn't remove duplicates in the destination. We could make it do that but that is not my idea here. We would need to add something like this when the source is hashed:
```
current_hashes = {}
for each row in source rows:
    digest = hasher(row).hexdigest()
    while digest in current_hashes:
        hasher.update(digest)
        digest = hasher.hexdigest()
    if digest not in past_hashes:
        # the row is new or updated
```
